File: computational_physics/Homework7/hw7_2.jl
using LinearAlgebra
using Plots

h = 0.05
k = 0.0025
x = 0:h:2
t = 0:k:0.2
boundary_conditions = [0, 0]
n = length(x)
m = length(t)
T = zeros(n, m)
T[1, :] .= boundary_conditions[1]
T[end, :] .= boundary_conditions[2]

# Initial conditions
for i in 1:n
    if 0 <= x[i] <= 1
        T[i, 1] = x[i]
    else
        T[i, 1] = -x[i] + 2
    end
end

factor = k / h^2

# Implicit method

A = diagm(0 => [1 + 2 * factor for i in 1:n-2], 1 => [-factor for i in 2:n-2], -1 => [-factor for i in 2:n-2])

for j in 2:m
    b = deepcopy(T[2:n-1, j - 1])
    b[1] = b[1]+factor * (T[1, j] )
    b[end] = b[end]+factor * (T[end, j] )
    solution = A \ b
    T[2:n-1, j] = solution
end


labels = ["Point $i" for i in t]
# display(T)
plot(x, T, color = :auto,label=labels,linewidth=3,
     xlabel = "distance [m]", ylabel = "Temperature [°C]", legend = false)
savefig("implicit_diffusion.png") 

File: computational_physics/Homework7/hw7_2.py
import numpy as np
import matplotlib.pyplot as plt
h=0.1
k=0.0025
x=np.arange(0,2+h,h)
t=np.arange(0,0.02+k,k)
boundary_conditions=[0,0]
n=len(x)
m=len(t)
T=np.zeros((n,m))
T[0,:]=boundary_conditions[0]
T[-1,:]=boundary_conditions[1]
#initial conditions
for i in range(n):
    if 0 <= x[i] <= 1:
        T[i] = x[i]
    else:
        T[i] = -x[i] + 2
print(T)
factor=k/h**2
# #implicit method
A=np.diag([1+2*factor]*(n-2),0)+np.diag([-factor]*(n-3),-1)+np.diag([-factor]*(n-3),+1)
for j in range(1,m):
    b=T[1:-1,j-1].copy()
    b[0]=b[0]+factor*T[0,j]
    b[-1]=b[-1]+factor*T[-1,j]
    solution=np.linalg.solve(A,b)
    print(solution)
    T[1:-1,j]=solution

print(T)
R=np.linspace(1,0,m)
B=np.linspace(0,1,m)
G=0
for j in range(m):
    plt.plot(x,T[:,j],color=[R[j],G,B[j]])
plt.xlabel('distance[m]')
plt.ylabel('Temperature[$\degree$ C]')
plt.legend([f't={value}s'for value in t.round(3)])
plt.show()
